pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -741,31 +730,46 @@ func (r *ReconcileClusterPool) createRandomNamespace(clp *hivev1.ClusterPool) (*
 	return ns, err
 }
 
+func getClustersToDelete(cds *cdCollection, deletionsNeeded int, logger log.FieldLogger) []*hivev1.ClusterDeployment {
+	installingClusters := cds.Installing()
+	if deletionsNeeded <= len(installingClusters) {
+		return installingClusters[:deletionsNeeded]
+	}
+
+	clustersToDelete := installingClusters
+	origDeletionsNeeded := deletionsNeeded
+	deletionsNeeded -= len(installingClusters)
+
+	standbyClusters := cds.Standby()
+	if deletionsNeeded <= len(standbyClusters) {
+		return append(clustersToDelete, standbyClusters[:deletionsNeeded]...)
+	}
+
+	clustersToDelete = append(clustersToDelete, standbyClusters...)
+	deletionsNeeded -= len(standbyClusters)
+
+	readyClusters := cds.Assignable()
+	if deletionsNeeded <= len(readyClusters) {
+		return append(clustersToDelete, readyClusters[:deletionsNeeded]...)
+	}
+
+	logger.WithField("deletionsNeeded", origDeletionsNeeded).
+		WithField("installingClusters", len(installingClusters)).
+		WithField("standbyClusters", len(standbyClusters)).
+		WithField("readyClusters", len(readyClusters)).
+		Error("trying to delete more clusters than there are available")
+
+	return append(clustersToDelete, readyClusters...)
```

Comment:
This would result in error eventually when we try to delete the cd according to the logs above - right? Shouldn't we fail early?

Reply:
No, it shouldn't result in an error. If we get here it just means that the length of the list we return will be _less_ than the number of deletions that were originally requested. The list will simply contain all the unassigned clusters (except broken ones -- they got deleted in a previous reconcile).